www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,290 @@
+locktag.locktag_field1 = logical_db_id;
+locktag.locktag_field2 = schema_name_hash;
+locktag.locktag_field3 = object_name_hash;
+
+res = LockAcquire(&locktag,AccessExclusiveLock…);
+```
+
+And just like we did for previous locks, we wrapped the code that acquires the user locks in a method that we then exposed as part of the `nile_ddl` extension. Before distributing `CREATE TABLE todos()` the transaction coordinator has to distribute `SELECT nile_ddl.nile_ddl_userlock(...)` which acquire the user locks which we hold for the table creation.
+
+In order to avoid lock contention, long blocking and deadlock we take few safety precautions in addition to the early lock acquisition:
+- We acquire locks in the same order on all databases
+- We always start acquiring locks with the same database, so there won't be distributed conflicts and local conflicts will be detected early
+- We set `LOCK_TIMEOUT` to a low value before executing anything related to distributed DDLs
+
+Even with all the precautions, things can still go wrong. In the next section we'll discuss our failure handling mechanisms.
+
+## When things go wrong
+
+Failure handling in distributed systems has to be built into every part of the architecture, it can't be added of as an afterthought. This is why we oriented our architecture around tried-and-true transaction handling - Postgres transaction mechanism and 2PC. The architecture provides the basic primitives for handling failure scenarios safely. In this section we'll look into key failure scenarios and how we used the components described above to deal with these scenarios.
+
+If you were to start listing out all the failure scenarios - the state of the database, the type of failures that can occure and the possible ways to handle each, you'll end up with a rather intimidating list. We went through this exercise and listed dozens of possible scenarios. This list is useful both to make sure we are thinking comprehensively, and also later on when it comes time to test our handling of all these scenarios.
+
+By examining all the failure scenarios, we derived two key insights that guided our design.
+
+The first insight is that there are two distinct types of failures: Failures that happen before the `PRECOMMIT` phase succeeds, which always result in `ROLLBACK` and failures that happen after and will always result in `COMMIT`.
+
+The transition happens when all databases respond to the `PREPARE TRANSACTION` command from the transaction coordinator.  We consider the responses to `PREPARE TRANSACTION` as votes from each database on the decision to commit. This to be a unanimous decision - if any database returns an error, we will force everyone to roll back with `ROLLBACK PREPARED` command. If every database responded successfully, we have a successful vote to commit, everyone is committed to commit the transaction. At this point, we say that the commit is **"doomed to succeed"**, as per Postgres we can no longer return errors or roll back.
```

Comment:
We can also mention that the transaction is 'doomed to succeed' not only  after successful vote to commit from remote databases, but after it commits on the originating database. If remote databases all voted to commit, but the originating database after that crashes, the transaction will be rolled back eventually.

Reply:
Thanks for mentioning that the originating DB doesn't participate in the 2PC. I'll leave this detail out of the blog for simplicity.